### PR TITLE
LLL: whole contract tests for ERC20 token.

### PR DESCRIPTION
**Pull request by benjaminion**
@axic this is in respect of #2447 as promised.

I've produced a test suite for an ERC20 contract I wrote in LLL. If this is along the lines you are looking for then let me know and I will expand the tests a bit more. Equally, if you have something else in mind please give me some guidance. It runs as follows:

```
test/soltest -t LLLERC20/* -- --ipcpath /tmp/testeth/geth.ipc
```

If this approach is OK then I'll also do @zigguratt's ENS Registry as a separate test suite in due course.

### CODE REVIEW

**Comment by axic** [comment]
@benjaminion this looks great, thanks! Will review it soon.

**Comment by benjaminion** [comment]
I've added a bunch more tests and tidied up a bit.

File: test/contracts/LLL_ERC20.cpp [hunk 1]
```diff
@@ -0,0 +1,590 @@
...
+      ;; ----------------------------------------------------------------------
+      ;; Fallback: No functions matched the function ID provided.
+
+      (revert)))
+  )
+)DELIMITER";
+
+static unique_ptr<bytes> s_compiledErc20;
+
+class LllErc20TestFramework: public LLLExecutionFramework
```
Comment (axic) [inline test/contracts/LLL_ERC20.cpp]: Please use `LLLERC20TestFramework` here.

File: test/contracts/LLL_ERC20.cpp [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+	BOOST_CHECK(callContractFunction("balanceOf(address)", ACCOUNT(0)) == encodeArgs(TOKENSUPPLY));
+}
+
+BOOST_AUTO_TEST_CASE(constants)
+{
+	deployErc20();
+
+	BOOST_CHECK(callContractFunction("totalSupply()") == encodeArgs(TOKENSUPPLY));
+	BOOST_CHECK(callContractFunction("decimals()") == encodeArgs(TOKENDECIMALS));
+	BOOST_CHECK(abiString(callContractFunction("symbol()")) == TOKENSYMBOL);
```
Comment (axic) [inline test/contracts/LLL_ERC20.cpp]: This should be encodeable via `encodeArgs` or `encodeDyn`:

```
        BOOST_CHECK(callContractFunction("symbol()") == encodeDyn(TOKENSYMBOL));
```

or

```
        string s(TOKENSYMBOL);
        bytes args = encodeArgs(u256(0x20), u256(s.length()), s);
        BOOST_CHECK(callContractFunction("symbol()") == args);
```

File: test/contracts/LLL_ERC20.cpp (hunk 1 above)
Comment (benjaminion) [reply test/contracts/LLL_ERC20.cpp]: Done.

File: test/contracts/LLL_ERC20.cpp (hunk 2 above)
Comment (benjaminion) [reply test/contracts/LLL_ERC20.cpp]: Great! That's better.  I had to do, `encodeDyn(string(TOKENSYMBOL))` for reasons not clear, but it works.

Just committing an update.

File: test/contracts/LLL_ERC20.cpp [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+	BOOST_CHECK(callContractFunction("balanceOf(address)", ACCOUNT(0)) == encodeArgs(TOKENSUPPLY));
+	BOOST_CHECK(callContractFunction("balanceOf(address)", ACCOUNT(1)) == encodeArgs(0));
+}
+
+BOOST_AUTO_TEST_CASE(invalid_transfer_2)
+{
+	deployErc20();
+
+	// Separate transfers that together exceed initial balance.
+	int transfer = 1 + TOKENSUPPLY/2;
```
Comment (axic) [inline test/contracts/LLL_ERC20.cpp]: Can you add some whitespace: `1 + TOKENSUPPLY / 2`?

File: test/contracts/LLL_ERC20.cpp [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+	m_sender = account(0);
+
+	// First transfer should succeed.
+	BOOST_REQUIRE(callContractFunction("transfer(address,uint256)", ACCOUNT(1), u256(transfer)) == SUCCESS);
+	BOOST_REQUIRE(callContractFunction("balanceOf(address)", ACCOUNT(0)) == encodeArgs(TOKENSUPPLY-transfer));
```
Comment (axic) [inline test/contracts/LLL_ERC20.cpp]: And here and below `TOKENSUPPLY - transfer`.

**Comment by axic** [comment]
Apart from the above two tiny things I think this is ready.

File: test/contracts/LLL_ERC20.cpp (hunk 3 above)
Comment (benjaminion) [reply test/contracts/LLL_ERC20.cpp]: Ok, though I don't like it :-)

File: test/contracts/LLL_ERC20.cpp (hunk 4 above)
Comment (benjaminion) [reply test/contracts/LLL_ERC20.cpp]: Done.

**Comment by benjaminion** [comment]
> Apart from the above two tiny things I think this is ready.

Actually, I've figured out how to access the event logging and have added some more tests around that, so I'm afraid there is a bit more to review... sorry!

File: test/contracts/LLL_ERC20.cpp [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+
+	// Transfer 1000 tokens from account(0) to account(1).
+	int transfer = 1000;
+	m_sender = account(0);
+	BOOST_REQUIRE(callContractFunction("transfer(address,uint256)", ACCOUNT(1), u256(transfer)) == SUCCESS);
+
+	// Check that a Transfer event was recorded and contents are correct.
+	BOOST_REQUIRE(m_logs.size() == 1);
+	BOOST_CHECK(m_logs[0].data == encodeArgs(transfer));
+	BOOST_CHECK(m_logs[0].topics[0] == keccak256(string("Transfer(address,address,uint256)")));
```
Comment (axic) [inline test/contracts/LLL_ERC20.cpp]: Should be there an assert for `topics.size() == 3`? Here and in the others.

File: test/contracts/LLL_ERC20.cpp [hunk 6]
```diff
@@ -0,0 +1,647 @@
...
+
+class LLLERC20TestFramework: public LLLExecutionFramework
+{
+protected:
+	void deployErc20()
+	{
+		if (!s_compiledErc20)
+		{
+			vector<string> errors;
+			s_compiledErc20.reset(new bytes(compileLLL(erc20Code, false, &errors)));
```
Comment (axic) [inline test/contracts/LLL_ERC20.cpp:399]: Maybe add a require so that `errors` is empty.

File: test/contracts/LLL_ERC20.cpp (hunk 5 above)
Comment (benjaminion) [reply test/contracts/LLL_ERC20.cpp]: Good idea. Done.

File: test/contracts/LLL_ERC20.cpp (hunk 6 above)
Comment (benjaminion) [reply test/contracts/LLL_ERC20.cpp:399]: Done.

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
Thanks, merged this now :)

**Pull request merged**